sdk/tests/conformance/programs/program-handling.html

```diff
@@ -63,56 +63,76 @@
-
-debug("Change fragment shader to one that will not link");
-compile(gl, fs, "fshader-not-link");
-debug("Draw orange");
-gl.uniform4fv(colorLoc, [1, 127/255, 0, 1]);
-wtu.drawUnitQuad(gl);
-wtu.checkCanvas(gl, [255, 127, 0, 255], "should be orange");
-
-debug("Try linking");
-gl.linkProgram(prg);
-assertMsg(gl.getProgramParameter(prg, gl.LINK_STATUS) == false, "link should fail");
-debug("Draw green to show even though link failed old program is still valid");
-gl.uniform4fv(colorLoc, [0, 1, 0, 1]);
-wtu.drawUnitQuad(gl);
-wtu.checkCanvas(gl, [0, 255, 0, 255], "should be green");
-
-debug("Detach and delete shaders");
-gl.detachShader(prg, vs);
-gl.detachShader(prg, fs);
-gl.deleteShader(vs);
-gl.deleteShader(fs);
-
-debug("Draw blue to show even though shaders are gone program is still valid");
-gl.uniform4fv(colorLoc, [0, 0, 1, 1]);
-wtu.drawUnitQuad(gl);
-wtu.checkCanvas(gl, [0, 0, 255, 255], "should be blue");
-
-debug("Call useProgram to show old program can not be made current again");
-gl.useProgram(prg);
-wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "should be invalid");
-
-debug("Draw purple to show the original program is still usable");
-gl.uniform4fv(colorLoc, [1, 0, 1, 1]);
-wtu.drawUnitQuad(gl);
-wtu.checkCanvas(gl, [255, 0, 255, 255], "should be purple");
-
-wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors");
+function testProgramInvalidation() {
+    debug('');
+    debug('== Testing invalidation of the current program ==');
+    var vs = wtu.loadShaderFromScript(gl, "vshader");
+    var fs = wtu.loadShaderFromScript(gl, "fshader");
+    var prg = wtu.createProgram(gl, vs, fs);
+    gl.useProgram(prg);
+    const positionLoc = gl.getAttribLocation(prg, 'a_position');
+    const colorLoc = gl.getUniformLocation(prg, 'u_color');
+
+    wtu.setupUnitQuad(gl, positionLoc);
+
+    debug("Draw red with valid program");
+    gl.uniform4fv(colorLoc, [1, 0, 0, 1]);
+    wtu.drawUnitQuad(gl);
+    wtu.checkCanvas(gl, [255, 0, 0, 255], "should be red");
+
+    debug("Change fragment shader to one that will not link");
+    compile(gl, fs, "fshader-not-link");
+    debug("Draw orange");
+    gl.uniform4fv(colorLoc, [1, 127/255, 0, 1]);
+    wtu.drawUnitQuad(gl);
+    wtu.checkCanvas(gl, [255, 127, 0, 255], "should be orange");
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors before relink");
+
+    debug("Try linking");
+    gl.linkProgram(prg);
+    assertMsg(gl.getProgramParameter(prg, gl.LINK_STATUS) == false, "link should fail");
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors after linkProgram");
+    debug("Attempt to draw green; because link failed, in WebGL, the draw should generate INVALID_OPERATION");
+    gl.uniform4fv(colorLoc, [0, 1, 0, 1]);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "because the current program has been invalidated, uniform* calls generate INVALID_OPERATION");
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors before draw");
```

Comment:
You probably don't need this second no-error check after just checking the error.
I think that if we suspect late or multiple stacked errors, we should modify glErrorShouldBe to abstract them away.

Reply:
Thanks for your feedback; I see your point. For the time being I'm going to leave in this second error check out of an abundance of caution, but will think about a follow-up PR to `WebGLTestUtils.glErrorShouldBe` that will capture all errors and ensure that there are none outside the passed set. Also please feel free to file an issue about that.